testutils/src/main/scala/za/co/absa/enceladus/testutils/datasetComparison/Flatten.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright 2018-2019 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.enceladus.testutils.datasetComparison
+
+import org.apache.log4j.{LogManager, Logger}
+import org.apache.spark.sql.{Column, DataFrame}
+import org.apache.spark.sql.functions.{expr, max}
+import org.apache.spark.sql.types._
+
+import scala.collection.mutable
+
+object Flatten {
+  /**
+    * Taken from AbsaOSS/Cobrix project spark-cobol utils
+    * @see https://github.com/AbsaOSS/cobrix
+    *
+    * Given an instance of [[DataFrame]] returns a list of columns for flattening.
+    * For how to use look at [[flattenDataFrame()]]
+    * All nested structures can be then flattened and arrays are projected as columns.
+    *
+    * Note. The method checks the maximum size for each array and that could perform slowly,
+    * especially on a vary big dataframes.
+    *
+    * @param df A dataframe
+    * @return A new dataframe with flat schema.
+    */
+  def flattenSchema(df: DataFrame): List[Column] = {
+    val logger: Logger = LogManager.getLogger("enceladus.testutils.Flattening")
+    val fields = new mutable.ListBuffer[Column]()
+    val stringFields = new mutable.ListBuffer[String]()
+
+    /**
+      * Aggregating arrays of primitives by projecting it's columns
+      *
+      * @param path path to an StructArray
+      * @param fieldNamePrefix Prefix for the field name
+      * @param structField StructField
+      * @param arrayType ArrayType
+      */
+    def flattenStructArray(path: String, fieldNamePrefix: String, structField: StructField, arrayType: ArrayType): Unit = {
+      val maxInd = df.agg(max(expr(s"size($path${structField.name})"))).collect()(0)(0).toString.toInt
+      var i = 0
```

Comment:
It looks like old Java way. 
You can use scalas "for loop with ranges".
It will look like: 
`for (i <- 0 to maxInd)
       doYourBusiness`
or `(0 to maxInd).map { i => doYourBusiness}`
or `(0 to maxInd).foreach { i => doYourBusiness}`
or `for{
  i <- (0 to maxInd)
} yield {
  doYourBusiness
}`

Reply:
While is used because it's the fastest loop.

Reply:
https://github.com/scala/bug/issues/1338 scala issue for that
Why ? http://ochafik.com/p_806